### ISSUE TITLE
Stop scanning autotest files if the code section of the last function has been obtained

### ISSUE THREAD

**Issue by peterwywong** [xcat2/xcat-core#6611]
subroutine get_files_recursive builds a file array by scanning directories recursively under /opt/xcat/share/xcat/tools/autotest/testcase. The resulting file array is scanned further to extract the code section for each test for subsequent testing.

For each file, it may have many sections of code. For example,

File_1 has
start test1_a
...
end
start test1_b
...
end

File_2 has
start test2_a
...
end
start test2_b
...
end

Let's use "xcattest -t test1_a,test2_b" for illustration purposes.

Files are searched sequentially based on the sequence they are searched by get_files_recursive. And the files are searched ONLY ONE TIME for code extraction.

When "start test_name" is scanned, test_name is matched against the list of tests.

For example, when "start test1_a" is scanned, xcattest knows that this is one of the tests, and the following code section is saved away for its testing later. And then keep on checking all "start" lines.
Later on, "start test2_b" is scanned and xcattest knows that this is one of the tests. The following code section is saved away similarly.

One thing I notice is that, after the second test is saved away, the current logic would scan the remaining files even though there are no more tests to match.

Therefore, I introduce a few variables to stop file scanning once the last test is saved away. Here is the idea:

$cases_left = $case_num;  # In this case, $cases_left = $case_num = 2.

$case_left is decremented by 1 whenever there is a match. When $case_left is 0, the optimization kicks in. We will leave the inner loop to stop scanning the remaining portion of the file. And we will also leave the outer loop to stop scanning remaining files.

This has been tested successfully for a list of tests and a bundle.

What happens if a test is tested more than one time (perhaps listed out mistakenly), say "xcatest -t test1_a,test1_a"?

The above change works correctly, but without early exit. It behaves the same as the existing code.

Let's go through the steps. In this case, $case_num is 2 initially for two identical tests.

Note that the file array is scanned ONLY ONE TIME, when test1_a is recognized, its code section is saved away and #cases_left is decremented to 1. However, we won't see test1_a again, so $case_left = 1 for the remaining of the file scanning. That is, no optimization is trigged. But the code section for test1_a is saved, so it will be tested twice.

The same thing is true if we have an invalid test, e.g., a typo in the test name. It is the same idea. $cases_left is never 0, so the optimization would not kick in. The new code would work correctly.

**Comment by gurevichmark** [comment]
@peterwywong What will happen with current code and with your changes if:
```
File_1 has
start test1_a
...
end
start test1_b
...
end

File_2 has
start test1_a
...
end
start test1_b
...
end
```
And then `xcattest -t test1_a,test1_b`

**Comment by peterwywong** [comment]
The code would work fine too and optimization should kick in in this case.

$cases_left = 2 initially.

When it first sees "start test1_a" and recognizes it as one of the tests, the code segment of test1_a is saved away. $cases_left = 1.

When it sees "start test1_b" and recognizes it as one of the tests, the code segment of test1_b is saved away. $cases_left =0. And then it exits early.

There is no need to scan File_2 since the code segments of test1_a and test1_b have been saved.

Without the optimization, I think the code segment of test1_a is saved again into the same data structure. It saves no new information.

**Comment by gurevichmark** [comment]
So sounds like with current, existing code, the **last** found testcases with names `test1_a` and `test1_b` will be executed. With your changes the **first** found testcases with names `test1_a` and `test1_b` will be executed ?

**Comment by peterwywong** [comment]
This is correct.

If they are different, the content of testcase is incorrect.

For example, lsdef_a is defined under lsdef and nowhere else. mkdef_f is defined under mkdef and nowhere else.
If lsdef_a is defined in two places and have different contents, then it should be corrected.

**Comment by gurevichmark** [comment]
Maybe the behavior should be different ? Maybe we should scan everything and report if duplicate testcase names are found ? Something like `Duplicate testcase name found: xxxxxx` ?

I am thinking of the case where a user adds their own testcase called `lsdef_a` and tries to run it. Depending on the name of their subdirectory, `xcattest` might run their testcase or ours.

Another solution is to update the man page for `xcattest` to recommend users run 
`xcattest -l caseinfo -t {testcasename}` to see if *testcasename* already exists ?

**Comment by peterwywong** [comment]
Based on the existing code, if a user adds his/her test named lsdef_a, xcattest may run the old lsdef_a or the new lsdef_a, depending on the sequence of files being decided by get_files_recursive. Indeed, we want a deterministic outcome.

As you said, we could provide a readme and tool (like your xcattest -l caseinfo -t {testcasename} above) to make sure there is no ambiguity on which one to be executed.

If there is an assumption that there is only one lsdef_a definition, my optimization should work.

If that assumption is not true, then we do need to modify the existing code to address that and this optimization does not take that into consideration.

**Comment by peterwywong** [comment]
xcattest was re-examined yesterday. In this investigation, multiple test cases had the same name. One of them used the os attribute with many values, such as linux, rhels and unknown. Casandra is correct that the os attribute is checked too.
A few key concepts need to be corrected. Some remain valid.
(1) The current test structure allows multiple test cases having the same name. For example, four test cases have the name XYZ. The LAST (not the FIRST) one will be run by Subroutine run_case, although all copies are saved away by Subroutine load_case. An index is associated with each copy of XYZ, and the "highest" index for XYZ is remembered and referenced when run_case is executed.

(2) In my earlier investigation, I did not have duplicates in mind. I had a wrong reading that once XYZ is found, XYZ is removed from the list of test cases to be run. In fact, the list of test cases to be run is unchanged throughout load_case. So, I incorrectly concluded that the FIRST occurrence of XYZ is used even there are other copies defined elsewhere.

(3) Subroutine get_files_recursive scans all files under testcase. One's expected test case may not be the last one, among many valid copies. Therefore, if a unique name is chosen for a new test case, it will be picked up without ambiguity. Indeed, all existing test case names must be examined first to avoid duplication.

(4) I found out duplicates this week. Some of them use "os" to distinguish themselves (e.g., rhels vs sles) and their contents are slightly different. Subroutine load_case has code to examine OSes. If the OS is invalid (e.g., centos), the remaining lines of the test case are scanned but not saved. If the OS is valid for the system it runs on (e.g., sles), it moves on to check "arch" similarly, if specified. If everything is fine, all the remaining lines of the test case are saved and its index becomes the highest index so far.

(5) Even we have seen 10 valid test cases for XYZ so far, the 11th copy may appear later. Therefore, my optimization to exit the scanning earlier does not work. This idea works if no duplicates are allowed. So, I will close Issue 6611: Stop scanning autotest files if the code section of the last function has been obtained and PR 6612: Exit file scanning earlier after the information of the last test is saved.
